### ISSUE TITLE
Thoughts on FITS Undefined card

### ISSUE THREAD

**Issue by timj** [astropy/astropy#8544]
Astropy understands FITS header cards with undefined values.  It assigns a card with a special value of `astropy.io.fits.card.Undefined`.   This is perfectly fine until you try to use the dict interface to an `astropy.io.fits.header` and don't know where the dict comes from.  The astro_metadata_translator interface is designed to work for any "header" that looks like a dict. This can mean Astropy headers, LSST PropertyList and AST FitsChan.  I'd like to write code that is like:

```python
if k in header and header[k] is not None:
    ...
```
but at the moment I have to special case this just for Astropy:

```python
if k in header and header[k] is not None and not isinstance(header[k], Undefined):
    ...
```

Would it be bad if `astropy.io.fits.header.__getitem__` returned `None` for an `Undefined` header and `__setitem__` converted `None` to `Undefined` internally?  This would make the dict interface look just like a normal Python thing and I don't think it's unreasonable for an undefined value in a FITS header to map to a Python `None`.

I can do a PR if there are no objections.

**Comment by saimn** [comment]
Hmm I don't know, on one hand I understand how it can be annoying to have to do a specific case for this, but on the other hand it seems that this has been there forever (I had a look at the pyfits repository), and I'm not sure that using `None` is enough to carry the meaning of an undefined card. Also currently setting a value to None is converted to an empty string.
Thoughts @MSeifert04 @drdavella ?

**Comment by timj** [comment]
from a "looks like a dict" point of view I can't really see why `None` doesn't fit for undefined value. It seems really close to the same meaning in Python and FITS.  Setting a value in a dict to None and having that turn up as an undefined FITS header card seems exactly what you want doesn't it?

There are very few tests relating to `Undefined` so I do agree that it's not entirely clear how many users of `astropy.io.fits` are using this support and whether anyone would notice if `__getitem__` started to return `None` (changing `__setitem__` to change a `None` to `Undefined` seems the right thing to do).

**Comment by timj** [comment]
fitsverify warning about an undefined value is a bit odd given that undefined values are part of the FITS standard. Why would it warn about something that is giving you the information? LSST and all my previous telescopes wrote in undefined values when the value was undefined (eg doing a dark so the airmass header is set to undefined). cfitsio and Starlink AST all understand how to parse these headers (as does astropy).

Python `None` has the same meaning as undefined in FITS (typeless and indicating that there is no value).

**Comment by saimn** [comment]
Reading again the standard, the `if any` could indeed suggest that this is valid, so you might well be right here. I'm almost convinced :), though still wondering why this was coded like this. 

![Capture d’écran de 2019-04-08 23-36-09](https://user-images.githubusercontent.com/311639/55758449-23c0c000-5a57-11e9-8fbf-a0372cf81fa3.png)

**Comment by timj** [comment]
I think you need to read the paragraph that follows the one you pasted:

> The value field, when present, shall contain the ASCII-text representation of a literal string constant, a logical constant, or a numerical constant, in the format specified in Sect. 4.2. The value field may be a null field; i.e., it may consist entirely of spaces, in which case the value associated with the keyword is undefined.

**Comment by mhvk** [comment]
It does seems logical to have "no value" be represented by `None`, so that one can be even shorter and do things like
```
value = header.get(k, None)
if value is not None:
    ...
```

**Comment by timj** [comment]
@mhvk that is exactly my use case.

**Comment by mhvk** [comment]
@timj - Yes, I didn't mean to do anything but show that in the even shorter form it really helped. 

@saimn - to me, it would seem to make sense to proceed with this. Indeed, it might be good to have it in 3.2 so that if we run into big problems it is easier to revert before the next LTS (4.0). I.e., I'd vote for taking @timj up on his offer!

**Comment by saimn** [comment]
Yes, I agree that the change seems reasonable. Let's go for it.

**Comment by timj** [comment]
Ok. I'll work on a patch.

**Comment by timj** [comment]
See #8572 for PR.

**Comment by bsipocz** [comment]
Addressed in #8572
